// @aem-sites/universal-editor-overlays/dist/NavigationHandler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.runNavigationHandler = void 0;

/*
 * Copyright 2022 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
var clickHandler = function clickHandler(_ref) {
  var event = _ref.event,
      editor = _ref.editor;
  var anchor = event.target.closest("A");

  if (anchor) {
    event.preventDefault();
    editor.navigateTo({
      href: anchor.href
    });
  }
};

var runNavigationHandler = function runNavigationHandler(_ref2) {
  var editor = _ref2.editor;
  document.addEventListener("click", function (event) {
    return clickHandler({
      event: event,
      editor: editor
    });
  }, {
    capture: true
  });
};

exports.runNavigationHandler = runNavigationHandler;
